demos/mantle_convection/adaptive_base_case/adaptive_base_case.py

```diff
@@ -0,0 +1,362 @@
+    # solutions at the previous mesh, to the current mesh.
+    T.interpolate(T_init)
+    u.interpolate(u_init)
+    p.interpolate(p_init)
+
+    t_adapt = TimestepAdaptor(delta_t, u, V, maximum_timestep=0.1, increase_tolerance=1.5)
+
+    # Stokes related constants (note that since these are included in UFL, they
+    # are wrapped inside Constant):
+    Ra = Constant(1e5)  # Rayleigh number
+    approximation = BoussinesqApproximation(Ra)
+
+    steady_state_tolerance = 1e-9
+
+    # Nullspaces and near-nullspaces:
+    Z_nullspace = create_stokes_nullspace(Z, closed=True, rotational=False)
+
+    # Next rename for output:
+    u.rename("Velocity")
+    p.rename("Pressure")
+
+    mesh.cartesian = True
+    boundary = get_boundary_ids(mesh)
+    gd = GeodynamicalDiagnostics(z, T, boundary.bottom, boundary.top)
+
+    temp_bcs = {
+        boundary.bottom: {'T': 1.0},
+        boundary.top: {'T': 0.0},
+    }
+
+    stokes_bcs = {
+        boundary.bottom: {'uy': 0},
+        boundary.top: {'uy': 0},
+        boundary.left: {'ux': 0},
+        boundary.right: {'ux': 0},
+    }
+
+    energy_solver = EnergySolver(T, u, approximation, delta_t, BackwardEuler, bcs=temp_bcs)
+    Told = energy_solver.T_old
+    Ttheta = 0.5*T + 0.5*Told
+    Told.assign(T)
+    stokes_solver = StokesSolver(z, approximation, Ttheta, bcs=stokes_bcs,
+                                 nullspace=Z_nullspace, transpose_nullspace=Z_nullspace,
+                                 solver_parameters="iterative")
+
+    # Now perform the time loop:
+    for ts in range(timestep, timestep+Nt):
+
+        # Write output:
+        if ts % output_frequency == 0:
+            output_file.write(u, p, T)
+
+        dt = t_adapt.update_timestep()
+        time += dt
+
+        # Solve Stokes sytem:
+        stokes_solver.solve()
+
+        # Temperature system:
+        energy_solver.solve()
+
+        # Compute diagnostics:
+        energy_conservation = abs(abs(gd.Nu_top()) - abs(gd.Nu_bottom()))
+
+        # Calculate L2-norm of change in temperature:
+        maxchange = sqrt(assemble((T - energy_solver.T_old)**2 * dx))
+
+        # Log diagnostics:
+        plog.log_str(f"{ts} {time} {float(delta_t)} {maxchange} "
+                     f"{gd.u_rms()} {gd.u_rms_top()} {gd.ux_max(boundary.top)} {gd.Nu_top()} "
+                     f"{gd.Nu_bottom()} {energy_conservation} {gd.T_avg()} ")
+
+        # Leave if steady-state has been achieved:
+        steady_state_reached = maxchange < steady_state_tolerance
+        if steady_state_reached:
+            log("Steady-state achieved -- exiting time-step loop")
+            break
+
+    return time, T, u, p, steady_state_reached
+
+
+# -
+
+# As explained above, we start with using the symbolic expressions for temperature, velocity,
+# and pressure. After the first first call to run_interval(), `T`, `u`, and `p` will
+# simply refer to the solution in the last timestep on the current mesh.
+
+T = T_init
+u = u_init
+p = p_init
```

Comment:
Are these `_init` variables needed at all outside of the function? Can we not just bring the definitions from lines 57-60 here?

Reply:
Yes, I think you're right. My initial approach was to do everything in the same order as in the base case, but now there's only very little left above the enormous function block, so I'll look at your suggestion of moving that up and merge the initialisation in here...